### PR TITLE
Update Totem T2 mapping, fix bug in VFAT footprint check, and implement suggestions from PR #41777

### PR DESCRIPTION
**Pull request by oljemark**
#### PR description:

For the TOTEM special foreseen on 25.6.2023, the new T2 telescope will be used (and removed after the run) to measure inelastic rate and veto diffractive p+p collisions, when measuring the elastic p+p cross section.
Previously, PR #41472 introduced the 2-channel unpacking for T2 and PR #41777 included some T2 DQM plots and fixes. As part of the approval for PR #41777 , some fixes and syntax safety improvements were asked for, see https://github.com/cms-sw/cmssw/pull/41777#issuecomment-1564189810 . This meant that the tracked T2 parameter useOlderT2TestFile was added to several python config files in EventFilter/CTPPSRawToDigi, and also some other cosmetic changes to those files. 
Last sunday @perrotta wrote a PR (#41784, #41785 ) to implement the asked-for changes, from which I include most changes into this PR.

Last week we also obtained new T2 test files with a new mapping (the final version), and a more diverse set of emulated T2 data frames. (Run 368080 had LE distributed around 5 and TE ~ 12 for all events, Run 368081 same LE value but TE ~9, and Run 368082 same as 368080 but with tracks in half the planes only). Testing the code with these new files exposed several more bugs in the T2 code in the common PPS unpacker.

On receiving more feedback from the detector experts, it turned out that one more member needed to be added to the T2Digi definition in DataFormats/TotemReco , to access header flags. @vavati also suggested that the DQM additions in PR #41777 needed to be expanded by adding the T2 as a DQMCalibrationSource, necessitating changes to RecoPPS/Configuration, also including the XML T2 geometry in a RecoPPS/Local T2 config file since the geometry was used in both the T2Rechit producer and T2 DQM module, but was not found in the database as far as we looked (thanks,  @grzanka !) . The multichannel unpacker from PR #41472 was also turned on by default, needing era modifiers for 2016-2018 PPS data taken without the T2 detector, for which the dummy mapping file mapping_totem_nt2_2021.xml caused an exception since it was copied from 2017 diamond mapping without the changes in T2 xml schema introduced in PR #41472.

At present the T2 CRC field in the VFAT frame in the test files are still just a constant stub, but in the next version of the T2 firmware expected within a few days this will be calculated, and can validate the present calculation in the code. The code is using the same CRC formula as the RP, but due to the inverted order of reading the VFAT frame between T2 & Roman Pots, this needed new helper functions, as did the VFAT signature 'A','C','E' header word intializer checked by the footprint helper.

#### PR validation:

A subset of  runTheMatrix tests relevant for the PPS detectors that might be affected by our code changes was run successfully after the era modifier fix noted above. Those were 136.793 & 1041 & 1042. The T2 unit test under RecoPPS/Local passed, while previously one unit test in  in CalibPPS/TimingCalibration & 25 DQM/Integration were seen to fail when run on lxplus due to the files being on tape, but succeed when cmsbot/Jenkins ran on locally cached copies, see this comment:  https://github.com/cms-sw/cmssw/pull/41785#issuecomment-1570004633

The PR #41785 still passed the tests even so ( https://github.com/cms-sw/cmssw/pull/41785#issuecomment-1566839728 ), although more changes have been added than is included there.

The T2 unit test in RecoPPS/Local was successfully run with the older definition & xml mapping, producing T2Digis, and likewise the totemt2_dqm_test_nino_cfg.py script on last week's new files, produced the expected DQM plots. The  totemt2_dqm_test_common_cfg.py script was run on PPS alignment run data with the other PPS detectors connected but not T2, showing that even though we include T2 in the common CTPPS RawToDigi and Reconstruction tasks, this does not negatively affect the other PPS detectors.

#### If this PR is a backport please specify the original PR and why you need to backport that PR. If this PR will be backported please specify to which release cycle the backport is meant for:

These three PR's, #41472 , #41777 and this one should be backported either to a special release on the 13_0_X cycle or the appropriate regular release on the 13_0_X . At a lower priority, backporting will also be done to 13_1_X.

### CODE REVIEW

File: RecoPPS/Local/test/totemT2NewDigi_reco_cfg.py [hunk 1]
```diff
@@ -32,13 +32,13 @@
...
 process.load('EventFilter.CTPPSRawToDigi.totemT2Digis_cfi')
 process.totemT2Digis.rawDataTag = cms.InputTag("rawDataCollector")
-process.totemDAQMappingESSourceXML.verbosity = cms.untracked.uint32(1)
-process.totemT2Digis.RawUnpacking.verbosity = cms.untracked.uint32(1)
-process.totemT2Digis.RawToDigi.verbosity = cms.untracked.uint32(3)
-process.totemT2Digis.RawToDigi.useOlderT2TestFile = cms.untracked.uint32(1)
-process.totemT2Digis.RawToDigi.printUnknownFrameSummary = cms.untracked.uint32(3)
-process.totemT2Digis.RawToDigi.printErrorSummary = cms.untracked.uint32(3)
-process.totemDAQMappingESSourceXML.multipleChannelsPerPayload = cms.untracked.bool(True)
+process.totemDAQMappingESSourceXML.verbosity = 1
```
Comment (grzanka) [inline RecoPPS/Local/test/totemT2NewDigi_reco_cfg.py:35]: @oljemark are you sure this is the only config file that needs to be adjusted this way ?

File: EventFilter/CTPPSRawToDigi/python/totemT2Digis_cfi.py [hunk 2]
```diff
@@ -3,10 +3,12 @@
 from EventFilter.CTPPSRawToDigi.totemVFATRawToDigi_cfi import totemVFATRawToDigi
 
 totemT2Digis = totemVFATRawToDigi.clone(
-    subSystem = cms.string('TotemT2'),
+    subSystem = 'TotemT2',
     RawToDigi = totemVFATRawToDigi.RawToDigi.clone(
-        testID = cms.uint32(0), #Some ID mismatch in test sample
-        testCRC = cms.uint32(0), # no need to test CRC for diamond frames
-        testECMostFrequent = cms.uint32(0) # show error in the DQM and then DAQ is sending resync, no need to test in the unpacker
+        testID = 0, #Some ID mismatch in test sample
```
Comment (grzanka) [inline EventFilter/CTPPSRawToDigi/python/totemT2Digis_cfi.py:8]: is it the only clone that needs to be adjusted ?

File: RecoPPS/Local/test/totemT2NewDigi_reco_cfg.py (hunk 1 above)
Comment (oljemark) [reply RecoPPS/Local/test/totemT2NewDigi_reco_cfg.py:35]: Hi, @grzanka . I think this kind of syntax improvement was a nice-to-have, while the ask-fors were making the olderT2 parameter tracked & that one & two print variables into bools. Doing code search for these three parameters they were all changed everywhere & reading the other T2 config files in this directory, I don't see much similar variable-changing statements (and skipping the multiyear old other config files that I don't know are in use anymore).

File: EventFilter/CTPPSRawToDigi/python/totemT2Digis_cfi.py (hunk 2 above)
Comment (oljemark) [reply EventFilter/CTPPSRawToDigi/python/totemT2Digis_cfi.py:8]: @grzanka Maybe yes, if the variable needs to be declared as a type at least once?

File: RecoPPS/Local/test/totemT2NewDigi_reco_cfg.py (hunk 1 above)
Comment (grzanka) [reply RecoPPS/Local/test/totemT2NewDigi_reco_cfg.py:35]: In my opinion its also nice to have, I just would like to be sure that the replacement was done in all places where needed

File: EventFilter/CTPPSRawToDigi/python/totemT2Digis_cfi.py (hunk 2 above)
Comment (oljemark) [reply EventFilter/CTPPSRawToDigi/python/totemT2Digis_cfi.py:8]: Hi again, @grzanka . The TotemTiming was the only one throwing an exception when I first added the tracked T2 parameter, without the present conditional read of the tracked parameter. Looking at the other clones, RP doesn't touch the RawToDigi PSet, and ctppsDiamond accessing it as a dict evidently doesn't overwrite the T2 parameter in the original.

**Comment by oljemark** [comment]
Hi, @grzanka . So if we want the T2 added to the calibration source that needs to be uncommented to be used,  should we still add the T2 to some ctpps-task around here: https://github.com/CTPPS/cmssw/blob/fo_nt2_mapfix2/RecoPPS/Configuration/python/recoCTPPS_cff.py#L16 ? And what about keeping the produced data files, in https://github.com/CTPPS/cmssw/blob/fo_nt2_mapfix2/RecoPPS/Configuration/python/RecoCTPPS_EventContent_cff.py ?

**Comment by grzanka** [comment]
> Hi, @grzanka . So if we want the T2 added to the calibration source that needs to be uncommented to be used, should we still add the T2 to some ctpps-task around here: https://github.com/CTPPS/cmssw/blob/fo_nt2_mapfix2/RecoPPS/Configuration/python/recoCTPPS_cff.py#L16 ? And what about keeping the produced data classes, in https://github.com/CTPPS/cmssw/blob/fo_nt2_mapfix2/RecoPPS/Configuration/python/RecoCTPPS_EventContent_cff.py ?

You need a DQM source https://github.com/CTPPS/cmssw/blob/fo_nt2_mapfix2/DQM/CTPPS/python/totemT2DQMSource_cfi.py
which expects `totemT2Digis` and `totemT2RecHits`

Somewhere in the execution chain you need to add modules which produce these objects. 

I believe adding necessary digitizer and rechit producer is needed in couple of places in the RecoPPS/Local, for example in python directory.

File: EventFilter/CTPPSRawToDigi/python/ctppsRawToDigi_cff.py [hunk 3]
```diff
@@ -132,22 +132,23 @@
...
 from EventFilter.CTPPSRawToDigi.ctppsPixelDigis_cfi import ctppsPixelDigis
-ctppsPixelDigis.inputLabel = cms.InputTag("rawDataCollector")
+ctppsPixelDigis.inputLabel = "rawDataCollector"
 
 from Configuration.Eras.Modifier_ctpps_2016_cff import ctpps_2016
 from Configuration.Eras.Modifier_ctpps_2017_cff import ctpps_2017
 from Configuration.Eras.Modifier_ctpps_2018_cff import ctpps_2018
 (ctpps_2016 | ctpps_2017 | ctpps_2018).toModify(ctppsPixelDigis, isRun3 = False )
 (ctpps_2016 | ctpps_2017 | ctpps_2018).toModify(totemDAQMappingESSourceXML_TotemTiming, sampicSubDetId = 6)
+(ctpps_2016 | ctpps_2017 | ctpps_2018).toModify(totemDAQMappingESSourceXML_TotemT2, multipleChannelsPerPayload = False)
```
Comment (grzanka) [inline EventFilter/CTPPSRawToDigi/python/ctppsRawToDigi_cff.py:151]: Add a line of comment explaining why the format has changed.

File: DQM/CTPPS/test/totemt2_dqm_test_nino_cfg.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+process.GlobalTag = GlobalTag(process.GlobalTag, '130X_dataRun3_HLT_v2', '')
+
+#Raw-to-digi
+process.load('EventFilter.CTPPSRawToDigi.ctppsRawToDigi_cff')
+
+# local RP reconstruction chain with standard settings
+process.load("RecoPPS.Configuration.recoCTPPS_cff")
+
+process.load('Geometry.ForwardCommonData.totemT22021V2XML_cfi')
+process.load('Geometry.ForwardGeometry.totemGeometryESModule_cfi')
```
Comment (grzanka) [inline DQM/CTPPS/test/totemt2_dqm_test_nino_cfg.py:56]:
```suggestion
```

Is it needed now ?

File: DQM/CTPPS/test/totemt2_dqm_test_nino_cfg.py [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+# CTPPS DQM modules
+process.load("DQM.CTPPS.ctppsDQM_cff")
+process.totemDAQMappingESSourceXML_TotemT2.verbosity = 1
+process.totemT2Digis.RawUnpacking.verbosity = 1
+process.totemT2Digis.RawToDigi.verbosity = 3
+process.totemT2Digis.RawToDigi.testCRC = 1
+process.totemT2Digis.RawToDigi.useOlderT2TestFile = False
+process.totemT2Digis.RawToDigi.printUnknownFrameSummary = True
+process.totemT2Digis.RawToDigi.printErrorSummary = True
+process.totemDAQMappingESSourceXML_TotemT2.multipleChannelsPerPayload = True
```
Comment (grzanka) [inline DQM/CTPPS/test/totemt2_dqm_test_nino_cfg.py:67]:
```suggestion
```

File: DQM/CTPPS/test/totemt2_dqm_test_common_cfg.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
+import FWCore.ParameterSet.Config as cms
+
+from Configuration.Eras.Era_Run3_cff import Run3
+from Configuration.Eras.Modifier_ctpps_cff import ctpps
```
Comment (grzanka) [inline DQM/CTPPS/test/totemt2_dqm_test_common_cfg.py:4]:
```suggestion
# needed to have ctppsDQMCalibrationSource properly working
from Configuration.Eras.Modifier_ctpps_cff import ctpps
```

**Pull request merged**